Improve the note stealing

paulfd opened the pull request:
So the previous note stealing was very conservative. Now, we make use of the average power computed by all voices after each block to steal the voice that had the lowest mean-squared score in the last 10 blocks, without further checks. The reasoning is that most of the time under such pressure a new note will always trump an old one, and we might as well take the lowest.

Also, I put the default polyphony to 256 and increased the max to 512.

Close #187

paulfd committed:
Set a default polyphony of 256

paulfd committed:
Memoize the mean power

paulfd committed:
Voice are unilateraly stolen now depending on their average power over the last 10 blocks.

paulfd committed:
Rename the "canBeStolen" function

jpcima commented on src/sfizz/HistoricalBuffer.h:
```diff
@@ -60,11 +62,18 @@ class HistoricalBuffer {
+        if (!validMean) {
+            mean = sfz::mean<ValueType>(buffer);
+            validMean = true;
+        }
```
You can track the power in more efficient ways which don't involve O(n) computation.

- one is always to keep the moving sum so you skip the O(n) completely.
  As a value gets pushed, remove from the moving sum the buffer's oldest value, and add the new one.
  A problem is that this can drift. A solution is to use fixed-point integers instead of floats, which make it exact.
  Then you can easily make the length way >16, which currently isn't really much.

- otherwise to track the signal level, another possibility commonly used:
  take the absolute values of input samples, pass it in lowpass filter that has different response times on rising and falling signal. It should be immediate on rising and slow on falling.
similar to this https://fr.wikipedia.org/wiki/Redresseur#Filtrage_de_la_tension_de_sortie_d'un_pont_de_diode

jpcima replied:
Ok my bad I misunderstood on first read.
Regardless the current solution faces a tiny problem, it's the power computation will be different depending how you render from the library, where in small buffers or large ones.
So the solutions above are relevant in this regard.

paulfd committed:
Updated tests

paulfd committed:
Cherry pick the OPF refactor from the smoothers

paulfd committed:
Changed the historical buffer to a smoothed envelope tracker

paulfd committed:
Add aging in the voice

paulfd committed:
Use age and average envelope for stealing

paulfd committed:
Do not change parameters unnecessarily

paulfd committed:
Corrected a false error message when changing the number of voices

paulfd committed:
Refine the stealing algorithm

paulfd commented:
I added the envelope/power tracker as a simple `abs` + super-lowpass filtering set at 5 Hz for now. We may further tweak this part, but at least it's independent of the block size now. I also added the voice age as a possible metric.

The heuristic now is to sort voices by age, set a threshold of the envelope (now set at "average envelope value over the voices / 4"), try to find a voice which fits the bill to kill. If we go beyond "max voice age / 2", we just pick the oldest one.

I also kill all "sister voices" now, which are started at the same time and with the same trigger. Writing this down I realize that this can be a problem if one of the sister voice is super strong, so we still need to refine all of this. 

Maybe consider the average envelope over all sister voices, or just do not kill sister voices?

paulfd commented:
The more I think about this the more I think we need an opcode so instrument makers can set a voice stealing that is adapted to their instrument behavior...

jpcima commented on src/sfizz/Voice.cpp:
```diff
@@ -718,3 +734,24 @@ void sfz::Voice::setupOscillatorUnison()
+        for (unsigned s = 0; s < buffer.getNumFrames(); ++s)
+            (*tempSpan)[s] = std::abs(buffer.getConstSpan(i)[s]);
+
+        channelPowerFilters[i].processLowpass(*tempSpan, *tempSpan);
```
Can the one-pole filter have:
- `inline float tickLowpass(float x)`
- `inline float tickHighpass(float x)`

There isn't benefit to run it in buffers, since it's data-dependent from one cycle to the next.
This eliminates the temp span.

paulfd replied:
Done :+1:

jpcima commented:
>Maybe consider the average envelope over all sister voices, or just do not kill sister voices?

Perhaps consider the max of all of them?
From the case of a loud voice like you mentioned, I think this to make the most sense.

paulfd committed:
Add a tick method to the OPF

paulfd committed:
Consider all sister voices when killing a note

paulfd committed:
Change the name of the envelope variables and filters

paulfd committed:
Consider the max of the sister voices

paulfd committed:
Add implicit conversions to AudioBuffer

paulfd committed:
Increase the buffer pool size

paulfd committed:
Add a helper to apply a gain span to effect inputs
Used for ramping out voice data

paulfd committed:
Killed voices pre-render their next block
This data is ramped out before the main rendering loop, to smooth the voice killings

paulfd commented:
> > Maybe consider the average envelope over all sister voices, or just do not kill sister voices?
> 
> Perhaps consider the max of all of them?
> From the case of a loud voice like you mentioned, I think this to make the most sense.

Did that, and I also added that killed voices pre-render their block in all relevant effect busses. This data is "ramped out" at the beginning of the block before the rendering of active voices. It's a linear ramp now, it could be an exponential ramp.

The behavior seems good, on hardcore piano stuff (like say Chopin or Liszt) with relatively low polyphony you don't "hear" dropped notes, and if you're way crazy on the polyphony you can sometimes distinguish something going wrong.

I was bitten by the fact that the voice kills itself btw. Maybe flagging ended voices and letting the synth clean them up is a better idea in the end.

~I should polish and split up the stealing function, it ended up quite large and convoluted.~

paulfd committed:
C++11 does not like constexpr

paulfd committed:
Move the voice rendering into a private method

paulfd committed:
Move the voice ordering as a free function

paulfd committed:
Move things around

jpcima commented:
Looking good. Another thing, you may consider as heuristic to check if there exists already other voices using the same instrument, and increase the killability score if this is the case.

paulfd commented:
I'll merge the current state and listed possible improvements for the next release based on our discussions. Thanks all!
